### ISSUE TITLE
Strict tracking mod sometimes allows to let go of the slider without penalty

### ISSUE THREAD

**Issue by Wieku** [ppy/osu#17482]
### Type

Game behaviour

### Bug description

As in the title. It seems to happen only when slider head is clicked early, but not always.

### Screenshots or videos

https://user-images.githubusercontent.com/5307179/160281980-3b0894a1-039c-4612-8cf0-a49d03f79ee0.mp4

### Version

Lazer 2022.327.0

### Logs

[performance-audio.log](https://github.com/ppy/osu/files/8357603/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/8357604/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/8357605/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/8357606/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/8357607/runtime.log)
[performance.log](https://github.com/ppy/osu/files/8357608/performance.log)

**Comment by bdach** [comment]
So as far as I can tell, this is a latent floating-point bug which reproduces in an intermittent manner. Steps to semi-reliably reproduce:

1. Download and import https://osu.ppy.sh/beatmapsets/537048#osu/1529759
2. Download, extract and import the following replay that I made:
[spaceman_atlas3 playing Erehamonika remixed by kors k - Der Wald (Kors K Remix) (Rucker) [Maze].zip](https://github.com/ppy/osu/files/8358155/spaceman_atlas3.playing.Erehamonika.remixed.by.kors.k.-.Der.Wald.Kors.K.Remix.Rucker.Maze.zip)
3. Play the replay over and over again until you notice an instance wherein the first slider tail isn't missed immediately upon dropping tracking.

As to why I'm saying this is a floating-point issue - I recommend trying the above procedure with a breakpoint set [on this line](https://github.com/ppy/osu/blob/8c5594b3eadd8afb6db153d0726eaa0b48359f02/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L576). When the bug happens, this line breakpoint will be hit.

The short explanation is that sometimes, depending on frame alignment and how the slider is hit precisely, the rewind block in `DrawableHitObject.Update()` can be hit in normal gameplay or forward replay playback. This is because of usage of `Result.TimeOffset + endTime`, which is imprecise. `ApplyResult()` sets `Result.TimeOffset` as follows:

https://github.com/ppy/osu/blob/8c5594b3eadd8afb6db153d0726eaa0b48359f02/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L672

This subtraction is _potentially lossy_ due to [catastrophic cancellation](https://en.wikipedia.org/wiki/Catastrophic_cancellation). Which is to say, after doing that subtraction, it is not guaranteed that

```
HitObject.GetEndTime() + (Time.Current - HitObject.GetEndTime()) == Time.Current
```

Whenever the left hand side of the above is larger due to the catastrophic cancellation, this bug is triggered.

As for the fix, I can offer two alternatives, and I'm not super sure which of them is better:

* [Solution 1](https://github.com/ppy/osu/compare/master...bdach:dho-store-absolute-time?expand=1) is rather immediate and just calls `Precision.IsDefinitelyBigger()` in the right place to alleviate this.
* [Solution 2](https://github.com/ppy/osu/compare/master...bdach:dho-store-absolute-time-2?expand=1) does some jiggling around so that `TimeAbsolute` can be stored raw, rather than being always computed based on offset. It's the more correct solution in a sense, but I don't like having to preserve the invariant of "only one of these 2 values must be populated".

  Inverting the current state of affairs (i.e. making `TimeOffset` be derived from `TimeAbsolute`) doesn't seem easy because there is no guarantee that defaults have been applied to the hitobject when the judgement result is being created, so the end time may not necessarily be valid for reading yet.

**Comment by peppy** [comment]
Is there a reason we can't drop the `TimeOffset` completely and use absolute values everywhere? In your second solution, the only remaining usage outside of tests seems to be to reset it back to zero.

**Comment by bdach** [comment]
Same reason as the one I gave for not swapping it around and making `TimeOffset` derived from `TimeAbsolute`. Currently `TimeAbsolute` is computed from the offset and the end time. But whether the end time is valid depends on when consumers query `TimeAbsolute`. Query it on a slider before default application and you'll get infinity back, or 0. This leads to objects just not appearing on the playfield.

I can try to fix that, such that judgements are constructed post default application and thus - in theory - can read the valid end time in the ctor, but it's a much riskier and more time consuming change.

**Comment by smoogipoo** [comment]
Sounds reasonable enough

**Comment by bdach** [comment]
Fixed by #22291.
